huaweicloud/services/cfw/resource_huaweicloud_cfw_firewall.go

```diff
@@ -0,0 +1,847 @@
+		}
+		return params
+	}
+	return nil
+}
+
+func buildCreateEastWestFirewallBodyParams(d *schema.ResourceData) map[string]interface{} {
+	bodyParams := map[string]interface{}{
+		"er_id":           utils.ValueIngoreEmpty(d.Get("east_west_firewall_er_id")),
+		"inspection_cidr": utils.ValueIngoreEmpty(d.Get("east_west_firewall_inspection_cidr")),
+		"mode":            utils.ValueIngoreEmpty(d.Get("east_west_firewall_mode")),
+	}
+	return bodyParams
+}
+
+func buildUpdateEastWestFirewallStatusBodyParams(d *schema.ResourceData, objectID string) map[string]interface{} {
+	bodyParams := map[string]interface{}{
+		"object_id": objectID,
+		"status":    d.Get("east_west_firewall_status"),
+	}
+	return bodyParams
+}
+
+func createFirewallWaitingForStateCompleted(ctx context.Context, d *schema.ResourceData, meta interface{}, t time.Duration) error {
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"PENDING"},
+		Target:  []string{"COMPLETED"},
+		Refresh: func() (interface{}, string, error) {
+			cfg := meta.(*config.Config)
+			region := cfg.GetRegion(d)
+			// createFirewallWaiting: Create a CFW firewall.
+			var (
+				createFirewallWaitingHttpUrl = "v3/{project_id}/jobs/{id}"
+				createFirewallWaitingProduct = "cfw"
+			)
+			createFirewallWaitingClient, err := cfg.NewServiceClient(createFirewallWaitingProduct, region)
+			if err != nil {
+				return nil, "ERROR", fmt.Errorf("error creating CFW client: %s", err)
+			}
+
+			createFirewallWaitingPath := createFirewallWaitingClient.Endpoint + createFirewallWaitingHttpUrl
+			createFirewallWaitingPath = strings.ReplaceAll(createFirewallWaitingPath, "{project_id}", createFirewallWaitingClient.ProjectID)
+			createFirewallWaitingPath = strings.ReplaceAll(createFirewallWaitingPath, "{id}", d.Id())
+
+			createFirewallWaitingOpt := golangsdk.RequestOpts{
+				KeepResponseBody: true,
+				OkCodes: []int{
+					200,
+				},
+				MoreHeaders: map[string]string{"Content-Type": "application/json"},
+			}
+
+			createFirewallWaitingResp, err := createFirewallWaitingClient.Request("GET", createFirewallWaitingPath, &createFirewallWaitingOpt)
+			if err != nil {
+				return nil, "ERROR", err
+			}
+
+			createFirewallWaitingRespBody, err := utils.FlattenResponse(createFirewallWaitingResp)
+			if err != nil {
+				return nil, "ERROR", err
+			}
+			statusRaw, err := jmespath.Search(`data.status`, createFirewallWaitingRespBody)
+			if err != nil {
+				return nil, "ERROR", fmt.Errorf("error parse %s from response body", `data.status`)
+			}
+
+			status := fmt.Sprintf("%v", statusRaw)
+
+			targetStatus := []string{
+				"Success",
+			}
+			if utils.StrSliceContains(targetStatus, status) {
+				return createFirewallWaitingRespBody, "COMPLETED", nil
+			}
+
+			pendingStatus := []string{
+				"Running",
+			}
+			if utils.StrSliceContains(pendingStatus, status) {
+				return createFirewallWaitingRespBody, "PENDING", nil
+			}
+
+			return createFirewallWaitingRespBody, status, nil
+		},
```

Comment:
I think this method can be optimized.
- The `client` can be passed in through parameters, no need to create a new one every time.
- Function `StrSliceContains` seems unnecessary.